### ISSUE TITLE
Wrong error message when *a* locale is missing on the server

### ISSUE THREAD

**Issue by vercas** [mobile-shell/mosh#1018]
On my desktop, I've got a bit of an unusual locale setting:  
```
LANG=en_GB.UTF-8
LANGUAGE=en_US:en_GB
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC=en_GB.UTF-8
LC_TIME=en_GB.UTF-8
LC_COLLATE=en_GB.UTF-8
LC_MONETARY=en_GB.UTF-8
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT=en_AU.UTF-8
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=
```
Note the `LC_MEASUREMENT` is en_**AU**.UTF-8, unlike the rest.  

This makes the mosh server error out:
```
The locale requested by LANG=en_GB.UTF-8 isn't available here.
Running `locale-gen en_GB.UTF-8' may be necessary.

The locale requested by LANG=en_GB.UTF-8 isn't available here.
Running `locale-gen en_GB.UTF-8' may be necessary.

mosh-server needs a UTF-8 native locale to run.

Unfortunately, the local environment (LANG=en_GB.UTF-8) specifies
the character set "US-ASCII",

The client-supplied environment (LANG=en_GB.UTF-8) specifies
the character set "US-ASCII".

locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=en_GB.UTF-8
LANGUAGE=en_US:en_GB
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC=en_GB.UTF-8
LC_TIME=en_GB.UTF-8
LC_COLLATE=en_GB.UTF-8
LC_MONETARY=en_GB.UTF-8
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT=en_AU.UTF-8
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=
Connection to SERVER closed.
/usr/bin/mosh: Did not find mosh server startup message. (Have you installed mosh on your server?)
```

It tells me to run `locale-gen en_GB.UTF-8` (which doesn't help) but what's actually missing is the Australian locale. After I added that, it works just fine.  
Setting `LC_ALL=en_GB.UTF-8` or `LC_MEASUREMENT=en_GB.UTF-8` also helped, but those were workarounds.  

The error message should blame the correct locale, and it also should say that the command needs to be run *on the server*.

**Comment by keithw** [comment]
I really appreciate the detailed writeup here! I think the main problem is that we really don't know exactly how to reverse-engineer what `locale-gen` command needs to be run on the server to make `setlocale( LC_ALL, "" )` succeed and produce a UTF-8 charset.

You can see the logic that mosh-server is using here: https://github.com/mobile-shell/mosh/blob/master/src/util/locale_utils.cc#L57-L68

This models the server's OS as only caring about LC_ALL, LC_CTYPE, and LANG, in that order, for purposes of choosing the charset. But clearly your server OS behaves differently. BTW: what OS is it?

In general I'm not sure we really know a better way to identify which environment variable is causing the problem or what the user has to run on the server to fix it. But, suggestions welcome.

**Comment by vercas** [comment]
@keithw Ubuntu Server 18.04 is on all my servers, and I can experienced the problem with all of them.  
I don't know the search logic either. One thing that can definitely be improved is the error message, though. It's not obvious that a certain locale is missing *from the server*.  
When I tried to get help over at #mosh @ Freenode, everyone told me to type that command on the client... So I'm not the only one getting confused.

**Comment by lebed2045** [comment]
same issue, it just doesn't work out of the box. Nor after some time when I tried to make it work. 
```mosh test-vps
The locale requested by LC_CTYPE=UTF-8 isn't available here.
Running `locale-gen UTF-8' may be necessary.

The locale requested by LC_CTYPE=UTF-8 isn't available here.
Running `locale-gen UTF-8' may be necessary.

mosh-server needs a UTF-8 native locale to run.

Unfortunately, the local environment (LC_CTYPE=UTF-8) specifies
the character set "US-ASCII",

The client-supplied environment (LC_CTYPE=UTF-8) specifies
the character set "US-ASCII".

locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=
LANGUAGE=
LC_CTYPE=UTF-8
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=
Connection to 165.227.18.118 closed.
/usr/local/bin/mosh: Did not find mosh server startup message. (Have you installed mosh on your server?)
```

1) I'm fairly confided I've install mosh on the server. 

```
test-vps:~$ mosh-server
MOSH CONNECT 60001 CSGKoKXx310WKuOUcgtTkA
mosh-server (mosh 1.3.2) [build mosh 1.3.2]
Copyright 2012 Keith Winstein <mosh-devel@mit.edu>
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

[mosh-server detached, pid = 3024]
```
2) about "The locale requested by LC_CTYPE=UTF-8 isn't available here.
Running `locale-gen UTF-8' may be necessary." When I'm trying to follow that advice,
```
test-vps:~$ locale-gen UTF-8
Error: 'UTF-8' is not a supported language or locale
``` 
- so this advice is contradictory. When I googled what might be a problem, I found this: https://askubuntu.com/questions/162391/how-do-i-fix-my-locale-issue
the long story short, it suggested to use `sudo locale-gen "en_US.UTF-8"` and I'm "will not see any errors anymore!"
but when I'm trying to run mosh again,

```
$ mosh test-vps
The locale requested by LC_CTYPE=UTF-8 isn't available here.
Running `locale-gen UTF-8' may be necessary.
```
so I see lots of errors again, which contradict suggestion form stackoverflow.

this is really frustrating experience ="(
is it possible to make mosh just work without playing mind games about what might work and what might not?

**Comment by keithw** [comment]
@lebed2045 In this case, your locale situation (between this client and this server) really does look to be screwed up, and the problem probably exists independent of Mosh. (It will be more subtle if you use SSH because SSH lets the problem through and doesn't warn -- Mosh is stricter and wants you to know that your settings are screwed up.) This sometimes happens because recent MacOS uses different locale names than Linux.

Try running something like: `ssh SERVERNAME perl` (no Mosh involved). You may also see a locale-related warning, e.g.:
```
$ LC_CTYPE=UTF-8 ssh localhost perl
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
	LANGUAGE = (unset),
	LC_ALL = (unset),
	LC_CTYPE = "UTF-8",
	LANG = "en_US.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to a fallback locale ("en_US.UTF-8").
```

The problem here is that your client is calling its locale `UTF-8` and your server doesn't have a locale by that name, and can't generate one. The server probably has one called `en_US.UTF-8` that it does know about. So your options include:

1. Tell the client the name of a locale that the server knows about. I.e., invoke Mosh or SSH with a prefix like this: `LC_ALL=en_US.UTF-8 mosh SERVERNAME` or `LC_ALL=en_US.UTF-8 ssh SERVERNAME`.

2. Figure out what on your client is setting the LC_CTYPE environment variable to UTF-8 and fix/change it. In most cases you don't want this variable set at all! Most configurations just have LANG=en_US.UTF-8 (assuming you want U.S. English with the UTF-8 charset) and none of the other locale-related environment variables.

See https://github.com/mobile-shell/mosh/issues/793 for more.

**Comment by ljcucc** [comment]
I tried but the solution is not working, is the config anything wrong?

local: `Darwin ljcucc-mbp.local 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:22 PDT 2022; root:xnu-8020.121.3~4/RELEASE_X86_64 x86_6`
```
$locale   
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
```

server: `Linux ljcuccpi 5.15.32-v8+ #1538 SMP PREEMPT Thu Mar 31 19:40:39 BST 2022 aarch64 GNU/Linux`
```
locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=en_GB.UTF-8
LANGUAGE=
LC_CTYPE=en_US.UTF-8
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_COLLATE="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT="en_GB.UTF-8"
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=
```

but with command:

```
LC_ALL=en_US.UTF-8 mosh ljcucc@ljcuccpi.local
ljcucc@ljcuccpi.local's password: 
The locale requested by LC_ALL=en_US.UTF-8 isn't available here.
Running `locale-gen en_US.UTF-8' may be necessary.

The locale requested by LC_ALL=en_US.UTF-8 isn't available here.
Running `locale-gen en_US.UTF-8' may be necessary.

mosh-server needs a UTF-8 native locale to run.

Unfortunately, the local environment (LC_ALL=en_US.UTF-8) specifies
the character set "US-ASCII",

The client-supplied environment (LC_ALL=en_US.UTF-8) specifies
the character set "US-ASCII".

locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=en_US.UTF-8
Connection to ljcuccpi.local closed.
/usr/local/bin/mosh: Did not find mosh server startup message. (Have you installed mosh on your server?)
```
